cylc/flow/job_file.py

```diff
@@ -128,6 +129,17 @@ def _write_header(handle, job_conf):
         """Write job script header."""
         handle.write("#!/bin/bash -l\n")
         handle.write("#\n# ++++ THIS IS A CYLC TASK JOB SCRIPT ++++")
+        # NOTE we cannot do /usr/bin/env bash because we need to use the -l
+        # option and GNU env doesn't support additional arguments (recent
+        # versions permit this with the -S option similar to BSD env but we
+        # cannot make the jump to this until is it more widely adopted
+        handle.write(dedent('''
+            if [[ $1 == 'noreinvoke' ]]; then
+                shift
+            else
+                exec "$(command -v bash)" -l "$0" noreinvoke "$@"
+            fi
+        '''))
```

Comment:
Yep @hjoliver that recap is correct.

I am not too keen on our use of login shells but I don't think this is a decision we can backtrack on at this stage.

After dwelling on the matter for a few days here's the solution I'm proposing. we invoke `/bin/bash` then immediately re-invoke using `$(command -v bash) -l`.

At first glance it looks pretty nasty but on reflection `/usr/bin/env bash` must surely re-invoke scripts using `exec` anyway so this is really just a lightweight `env` implementation which only does the thing we need it to do.

It relies on a system bash being installed as `/bin/bash`,  this is pretty reliable, if we are really going for it we could change that to `/bin/sh` which would support dash on some overly paired down system the likes of which I'm yet to encounter [insert bashism rant].

The full solution to this and other job script / bash woes is probably to reconsider the whole job script approach and possibly convert the whole thing to Python but that's a topic for another day.